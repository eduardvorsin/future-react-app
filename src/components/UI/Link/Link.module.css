.link {
	display: inline-flex;
	align-items: center;
	border-radius: 4px;
	font-size: 14px;
	color: var(--brand-color-200);
	border-width: 2px;
	border-style: solid;
	border-color: var(--brand-color-200);
	padding: 8px;
	text-decoration: none;
	transition:
		background-color var(--transition-time) var(--transition-easing),
		border var(--transition-time) var(--transition-easing),
		color var(--transition-time) var(--transition-easing),
		transform var(--transition-time) var(--transition-easing),
		opacity var(--transition-time) var(--transition-easing);
}

.link--primary {
	background-color: transparent;
}

.link--secondary {
	background-color: var(--brand-color-200);
	color: var(--dark-color-100);
	text-align: center;
	font-weight: 600;
}

.link--tertiary {
	transform: translateY(0);
	background: var(--brand-color-100);
	color: var(--secondary-color-100);
	border-color: var(--brand-color-100);
	line-height: 1.5;
	font-size: 16px;
	min-width: 130px;
	justify-content: center;
	padding: 7px 15px;
	font-weight: 700;
}

.link__icon-wrapper {
	pointer-events: none;
	width: 24px;
	height: 24px;
	fill: currentcolor;
}

.link__icon-wrapper--left {
	margin-right: 6px;
}

.link__icon-wrapper--right {
	margin-left: 6px;
}

.link--disabled {
	cursor: not-allowed;
	transition:
		color var(--transition-time) var(--transition-easing),
		border-color var(--transition-time) var(--transition-easing),
		background-color var(--transition-time) var(--transition-easing);
}

.link--primary:hover:not(.link--disabled) {
	background-color: var(--secondary-color-200);
	transition:
		background-color var(--transition-time) var(--transition-easing);
}

.link--secondary:hover:not(.link--disabled) {
	background-color: var(--secondary-color-200);
	transition:
		background-color var(--transition-time) var(--transition-easing);
}

.link--tertiary:hover:not(.link--disabled) {
	background-color: var(--secondary-color-100);
	transform: translateY(2px);
	color: var(--brand-color-100);
	transition:
		background-color var(--transition-time) var(--transition-easing),
		color var(--transition-time) var(--transition-easing),
		transform var(--transition-time) var(--transition-easing);
}

@media(prefers-color-scheme:dark) {
	.link--primary:hover:not(.link--disabled) {
		background-color: var(--secondary-color-300);
		transition:
			background-color var(--transition-time) var(--transition-easing);
	}

	.link--secondary:hover:not(.link--disabled) {
		background-color: var(--secondary-color-200);
		color: var(--light-color-300);
		transition:
			background-color var(--transition-time) var(--transition-easing),
			color var(--transition-time) var(--transition-easing);
	}

	.link--disabled {
		opacity: 0.5;
		transition:
			opacity var(--transition-time) var(--transition-easing);
	}
}

@media(prefers-color-scheme:light) {
	.link--disabled {
		color: var(--disabled-color-200);
		border-color: var(--disabled-color-200);
		background-color: var(--disabled-color-100);
		transition:
			color var(--transition-time) var(--transition-easing),
			border-color var(--transition-time) var(--transition-easing),
			background-color var(--transition-time) var(--transition-easing);
	}
}