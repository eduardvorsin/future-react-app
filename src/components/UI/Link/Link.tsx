import React, { FC, Fragment, FunctionComponent, MouseEventHandler, ReactNode } from 'react';
import { Link as RouterLink } from 'react-router-dom';
import classes from './Link.module.css';
import { SVGAndHTMLProps } from '../../../types/shared';

type BaseLinkProps = Partial<React.AnchorHTMLAttributes<HTMLAnchorElement>>;

interface ButtonProps extends BaseLinkProps {
  children: ReactNode,
  path?: string,
  className?: string,
  onClick?: MouseEventHandler<HTMLAnchorElement>,
  icon?: FunctionComponent<SVGAndHTMLProps>,
  iconPosition?: 'left' | 'right',
  isRouterLink?: boolean,
  variant?: 'primary' | 'secondary' | 'tertiary',
}

const Link: FC<ButtonProps> = ({
  path,
  className,
  children,
  onClick,
  icon,
  isRouterLink = false,
  iconPosition = 'left',
  variant = 'primary',
  ...props
}) => {
  const linkClasses = [
    classes.link,
    className,
    !path ? classes['link--disabled'] : '',
    classes[`link--${variant}`],
  ].join(' ');

  const iconClasses = [
    classes['link__icon-wrapper'],
    iconPosition === 'left' ? classes['link__icon-wrapper--left'] : classes['link__icon-wrapper--right'],
  ].join(' ');

  const Icon = icon ?? Fragment;

  const IconWrapper: JSX.Element = (
    <span
      className={iconClasses}
      aria-hidden='true'
    >
      <Icon />
    </span>
  );

  const Children: JSX.Element = (
    <>
      {(icon && iconPosition === 'left') && IconWrapper}
      {children}
      {(icon && iconPosition === 'right') && IconWrapper}
    </>
  );

  if (isRouterLink) {
    return (
      <RouterLink
        className={linkClasses}
        to={path ?? ''}
        onClick={onClick}
        aria-disabled={!path}
        {...props}
      >
        {Children}
      </RouterLink>
    );
  }

  return (
    <a
      className={linkClasses}
      href={path}
      rel="noopener"
      onClick={onClick}
      aria-disabled={!path}
      role={!path ? 'link' : undefined}
      {...props}
    >
      {Children}
    </a>
  );
};

export default Link;
