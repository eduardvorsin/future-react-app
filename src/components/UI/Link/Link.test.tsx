import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import React, { MouseEvent } from 'react';
import { RouterProvider, createMemoryRouter } from 'react-router-dom';
import Link from './Link';
import SunIcon from '../../../assets/images/icons/sun.svg';

describe('Link tests', () => {
  it('is rendered correctly', () => {
    render(
      <Link
        path='/'
      >
        test link
      </Link>,
    );

    expect(screen.getByRole<HTMLAnchorElement>('link')).toBeInTheDocument();
    expect(screen.getByRole<HTMLAnchorElement>('link').rel).toBe('noopener');
  });
  it('If the path prop is empty or null/undefined, then the link-disabled class is added', () => {
    render(
      <Link
      >
        test link
      </Link>,
    );

    expect(screen.getByRole<HTMLAnchorElement>('link')).toHaveClass('link--disabled');
  });
  it('when you click on the link, the mock function is triggered', async () => {
    const user = userEvent.setup();
    const mockFn = jest.fn<void, [e: MouseEvent]>();

    render(
      <Link
        path='/'
        onClick={mockFn}
      >
        test link
      </Link>,
    );

    await user.click(screen.getByRole<HTMLAnchorElement>('link'));

    expect(mockFn).toHaveBeenCalledTimes(1);
  });

  it('if the icon prop is passed, the icon will be rendered inside the link', () => {
    render(
      <Link
        path='/'
        icon={<SunIcon data-testid='test-icon' />}
      >
        test link
      </Link>,
    );

    expect(screen.getByTestId<HTMLElement>('test-icon')).toBeInTheDocument();
  });

  it('if the icon prop is passed, then the icon is rendered and the wrapper has the link__icon-wrapper--left class', () => {
    render(
      <Link
        path='/'
        icon={<SunIcon />}
      >
        test link
      </Link>,
    );

    expect(screen.getByTestId<HTMLSpanElement>('icon-wrapper')).toHaveClass('link__icon-wrapper--left');
  });
  it('if the icon prop is passed and the icon Position prop is equal to right, then the wrapper has the link__icon-wrapper--right class', () => {
    render(
      <Link
        path='/'
        icon={<SunIcon />}
        iconPosition='right'
      >
        test link
      </Link>,
    );

    expect(screen.getByTestId<HTMLSpanElement>('icon-wrapper')).toHaveClass('link__icon-wrapper--right');
  });

  it('if the isRouterLink prop is passed, then the link for the router is rendered', () => {
    const router = createMemoryRouter([{
      path: '/',
      element: (
        <Link
          path='/'
          isRouterLink
        >
          test link
        </Link>
      ),
    }], {
      initialEntries: ['/'],
    });
    render(<RouterProvider router={router} />);

    expect(screen.getByRole<HTMLAnchorElement>('link')).toBeInTheDocument();
    expect(screen.getByRole<HTMLAnchorElement>('link')).not.toHaveAttribute('rel');
  });

  it('is a basic snapshot', () => {
    render(
      <Link
        path='/'
      >
        test link
      </Link>,
    );

    expect(screen.getByRole<HTMLAnchorElement>('link')).toMatchSnapshot();
  });

  it('this is a snapshot with an icon on the left', () => {
    render(
      <Link
        path='/'
        icon={<SunIcon />}
      >
        test link
      </Link>,
    );

    expect(screen.getByRole<HTMLAnchorElement>('link')).toMatchSnapshot();
  });

  it('this is a snapshot with an icon on the right', () => {
    render(
      <Link
        path='/'
        icon={<SunIcon />}
      >
        test link
      </Link>,
    );

    expect(screen.getByRole<HTMLAnchorElement>('link')).toMatchSnapshot();
  });

  it('is a snapshot with a link from the router', () => {
    const router = createMemoryRouter([{
      path: '/',
      element: (
        <Link
          path='/'
          isRouterLink
        >
          test link
        </Link>
      ),
    }], {
      initialEntries: ['/'],
    });
    render(<RouterProvider router={router} />);

    expect(screen.getByRole<HTMLAnchorElement>('link')).toMatchSnapshot();
  });
});
