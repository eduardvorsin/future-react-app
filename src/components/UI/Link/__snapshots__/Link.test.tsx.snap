// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Link tests is a basic snapshot 1`] = `
<a
  aria-disabled="false"
  class="link   link--primary"
  href="/"
  rel="noopener"
>
  test link
</a>
`;

exports[`Link tests is a snapshot with a link from the router 1`] = `
<a
  aria-disabled="false"
  class="link   link--primary"
  href="/"
>
  test link
</a>
`;

exports[`Link tests this is a snapshot with an icon on the left 1`] = `
<a
  aria-disabled="false"
  class="link   link--primary"
  href="/"
  rel="noopener"
>
  <span
    aria-hidden="true"
    class="link__icon-wrapper link__icon-wrapper--left"
    data-testid="icon-wrapper"
  >
    <svg />
  </span>
  test link
</a>
`;

exports[`Link tests this is a snapshot with an icon on the right 1`] = `
<a
  aria-disabled="false"
  class="link   link--primary"
  href="/"
  rel="noopener"
>
  <span
    aria-hidden="true"
    class="link__icon-wrapper link__icon-wrapper--left"
    data-testid="icon-wrapper"
  >
    <svg />
  </span>
  test link
</a>
`;
